File: app/src/main/java/com/github/luecy1/nhkbangumi/task/MainAsyncTask.java
package com.github.luecy1.nhkbangumi.task;

import android.os.AsyncTask;
import android.util.Log;

import com.github.luecy1.nhkbangumi.Loading;
import com.github.luecy1.nhkbangumi.MainActivityListAdapter;
import com.github.luecy1.nhkbangumi.entity.nowonair.NowOnAir;
import com.github.luecy1.nhkbangumi.entity.nowonair.NowOnAirRoot;
import com.google.gson.Gson;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * Created by you on 2018/01/08.
 */

public class MainAsyncTask extends AsyncTask<Integer, Integer, List<NowOnAir>> {

    private String url;
    private MainActivityListAdapter adapter;
    private Loading loading;

    public MainAsyncTask(String url, MainActivityListAdapter adapter, Loading loading) {
        this.url = url;
        this.adapter = adapter;
        this.loading = loading;
    }

    @Override
    protected List<NowOnAir> doInBackground(Integer... integers) {

        Request req = new Request
                .Builder()
                .url(url)
                .build();

        OkHttpClient client = new OkHttpClient();

        NowOnAirRoot json = null;

        try {
            Response resp = client.newCall(req).execute();

            Gson gson = new Gson();
            json = gson.fromJson(resp.body().string(), NowOnAirRoot.class);


            resp.body().close();

        } catch (IOException e) {
            Log.e("MyApp", "IOExceptionが発生しました。");
            e.printStackTrace();
        }

        List<NowOnAir> programs = new ArrayList<>();

        if (json == null || json.getNowOnAirList() == null) {
            return programs;
        }

        // TV
        ifAddNonNull(programs, json.getNowOnAirList().getG1());
        ifAddNonNull(programs, json.getNowOnAirList().getG2());
        ifAddNonNull(programs, json.getNowOnAirList().getE1());
        ifAddNonNull(programs, json.getNowOnAirList().getE2());
        ifAddNonNull(programs, json.getNowOnAirList().getE3());
        ifAddNonNull(programs, json.getNowOnAirList().getE4());
        ifAddNonNull(programs, json.getNowOnAirList().getS1());
        ifAddNonNull(programs, json.getNowOnAirList().getS2());
        ifAddNonNull(programs, json.getNowOnAirList().getS3());
        ifAddNonNull(programs, json.getNowOnAirList().getS4());

        // radio
        ifAddNonNull(programs, json.getNowOnAirList().getR1());
        ifAddNonNull(programs, json.getNowOnAirList().getR2());
        ifAddNonNull(programs, json.getNowOnAirList().getR3());

        // netradio
        ifAddNonNull(programs, json.getNowOnAirList().getN1());
        ifAddNonNull(programs, json.getNowOnAirList().getN2());
        ifAddNonNull(programs, json.getNowOnAirList().getN3());

        return programs;
    }

    private void ifAddNonNull(List<NowOnAir> programs, NowOnAir nowOnAir) {
        if (nowOnAir != null) {
            programs.add(nowOnAir);
        }
    }

    @Override
    protected void onPostExecute(List<NowOnAir> programList) {

        if (programList != null) {
            adapter.addProgramList(programList);
        }

        loading.close();
    }
}

File: app/src/main/java/com/github/luecy1/nhkbangumi/task/MainAsyncTask.kt
package com.github.luecy1.nhkbangumi.task

import android.os.AsyncTask
import android.util.Log
import com.github.luecy1.nhkbangumi.Loading
import com.github.luecy1.nhkbangumi.MainActivityListAdapter
import com.github.luecy1.nhkbangumi.entity.nowonair.NowOnAir
import com.github.luecy1.nhkbangumi.entity.nowonair.NowOnAirRoot
import com.google.gson.Gson
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.*

/**
 * Created by you on 2018/01/08.
 */

class MainAsyncTask(private val url: String, private val adapter: MainActivityListAdapter, private val loading: Loading) : AsyncTask<Int, Int, List<NowOnAir>>() {

    override fun doInBackground(vararg params: Int?): List<NowOnAir>? {

        val req = Request.Builder()
                .url(url)
                .build()

        val client = OkHttpClient()

        var json: NowOnAirRoot? = null

        try {
            val resp = client.newCall(req).execute()

            val gson = Gson()
            json = gson.fromJson(resp.body()!!.string(), NowOnAirRoot::class.java)


            resp.body()!!.close()

        } catch (e: IOException) {
            Log.e("MyApp", "IOExceptionが発生しました。")
            e.printStackTrace()
        }

        val programs = ArrayList<NowOnAir>()

        if ((json?.nowOnAirList != null).not()) {
            return programs
        }

        // TV
        ifAddNonNull(programs, json?.nowOnAirList!!.g1)
        ifAddNonNull(programs, json.nowOnAirList!!.g2)
        ifAddNonNull(programs, json.nowOnAirList!!.e1)
        ifAddNonNull(programs, json.nowOnAirList!!.e2)
        ifAddNonNull(programs, json.nowOnAirList!!.e3)
        ifAddNonNull(programs, json.nowOnAirList!!.e4)
        ifAddNonNull(programs, json.nowOnAirList!!.s1)
        ifAddNonNull(programs, json.nowOnAirList!!.s2)
        ifAddNonNull(programs, json.nowOnAirList!!.s3)
        ifAddNonNull(programs, json.nowOnAirList!!.s4)

        // radio
        ifAddNonNull(programs, json.nowOnAirList!!.r1)
        ifAddNonNull(programs, json.nowOnAirList!!.r2)
        ifAddNonNull(programs, json.nowOnAirList!!.r3)

        // netradio
        ifAddNonNull(programs, json.nowOnAirList!!.n1)
        ifAddNonNull(programs, json.nowOnAirList!!.n2)
        ifAddNonNull(programs, json.nowOnAirList!!.n3)

        return programs
    }

    private fun ifAddNonNull(programs: MutableList<NowOnAir>, nowOnAir: NowOnAir?) {
        if (nowOnAir != null) {
            programs.add(nowOnAir)
        }
    }

    override fun onPostExecute(programList: List<NowOnAir>?) {

        if (programList != null) {
            adapter.addProgramList(programList)
        }

        loading.close()
    }
}
